README.md
---

## uxcore-formatter [![Dependency Status](http://img.shields.io/david/uxcore/uxcore-formatter.svg?style=flat-square)](https://david-dm.org/uxcore/uxcore-formatter) [![devDependency Status](http://img.shields.io/david/dev/uxcore/uxcore-formatter.svg?style=flat-square)](https://david-dm.org/uxcore/uxcore-formatter#info=devDependencies) 

## TL;DR

Formatter 是一些处理字符串的工具函数的集合，他的设计原则是第一个参数为一个 string，返回值为一个 string，后面的参数皆为可选，即有默认值。

#### setup develop environment

```sh
$ git clone https://github.com/uxcore/uxcore-formatter
$ cd uxcore-formatter
$ npm install
$ npm start
```

## Usage
```
var Formatter = require("uxcore-formatter");
console.log(Formatter.date(new Date(), 'YYYY-MM-DD');
```

## API
> 目前提供的函数包括：

### Date
```javascript
Formatter.date(new Date(), 'YYYY-MM-DD HH:mm:ss'); // 2015-10-12 14:22:16
```

### money

```javascript
Formatter.money("6566456.65466545") // "6 566 456.65466545"
Formatter.money("6566456.65466545", ",")  // "6,566,456.65466545"
Formatter.money("6566456", ",")  // "6,566,456"
Formatter.money("6566456", ",", 4) // "6,566,456.0000"
Formatter.money("6566456", ",", 2) // "6,566,456.00"
Formatter.money("6566456.65466545", ",", 4)  // "6,566,456.6547"
Formatter.money("6566456.65466545", ",", 2) // "6,566,456.65"
```

### cnmobile
```javascript
Formatter.cnmobile('+8615652988282', ' '); // +86 1565 2988 282
```

### card
```javascript
Formatter.card('1565298828212233', ' '); // 1565 2988 2821 2233
```




HISTORY.md
## 0.1.12
* `FIXED` fix bug when fixNum is 0

## 0.1.11

* `CHANGED` remove useless dependency `object-assign`, `prop-types`

# 0.1.9
* `CHANGED` use `big.js` instead of `parseFloat` to handle large number fixed

# 0.1.8
* `CHANGED` undefined/null/''  return ''
* because they are all invalid value for date.

# 0.1.7
* `CHANGED` null check

# 0.1.6
* `CHANGED` fit React@15
* `CHANGED` delete 'package-lock.json' file
# 0.1.3

* `CHANGED` eslint
* `CHANGED` add tests
* `CHANGED` use default delimeter only delimeter is undefined

# 0.1.2

`CHANGED` remove es6 syntax

# 0.1.1

`CHANGED` more rational money formatter


tests/Formatter.spec.js
import expect from 'expect.js';
import Formatter from '../src';

String.prototype.padStart = function (targetLength, padString) {
  padString = padString || ' ';
  if (this.length >= targetLength) {
    return this;
  }
  const gap = targetLength - this.length;
  const times = Math.floor(gap / padString.length);
  let result = '';
  for (let i = 0; i < times; i++) {
    result += padString;
  }
  return result + padString.slice(0, gap % padString.length) + this;
};

describe('Formatter.date', () => {
  const date = new Date();
  it('returns empty string when date is invalid', () => {
    expect(Formatter.date('test!@#$%^', 'YYYY-MM-DD')).to.be('');
  });

  it('returns empty string when date is null', () => {
    expect(Formatter.date(null, 'YYYY-MM-DD')).to.be('');
    expect(Formatter.date(undefined, 'YYYY-MM-DD')).to.be('');
    expect(Formatter.date(false, 'YYYY-MM-DD')).to.be('');
    expect(Formatter.date('', 'YYYY-MM-DD')).to.be('');
  });

  it('works fine', () => {
    expect(Formatter.date(date, 'YYYY-MM-DD')).to.be(`${date.getFullYear()}-${`${date.getMonth() + 1}`.padStart(2, '0')}-${`${date.getDate()}`.padStart(2, '0')}`);
  });

  it('works without pattern', () => {
    expect(Formatter.date(date)).to.be(`${date.getFullYear()}-${`${date.getMonth() + 1}`.padStart(2, '0')}-${`${date.getDate()}`.padStart(2, '0')}`);
  });

  it('works with custom pattern', () => {
    expect(Formatter.date(date, 'MM-DD')).to.be(`${date.getMonth() < 9 ? `0${date.getMonth() + 1}` : date.getMonth() + 1}-${date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()}`);
    expect(Formatter.date(date, 'DD')).to.be(`${date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()}`);
    expect(Formatter.date(date, 'MM-dd')).to.be(`${date.getMonth() < 9 ? `0${date.getMonth() + 1}` : date.getMonth() + 1}-${date.getDate() < 10 ? `0${date.getDate()}` : date.getDate()}`);
    expect(Formatter.date(date, 'QQ')).to.be(`${Math.floor(date.getMonth() / 3) + 1}`.padStart(2, '0'));
    expect(Formatter.date(date, 'qq')).to.be(`${Math.floor(date.getMonth() / 3) + 1}`.padStart(2, '0'));
    expect(Formatter.date(date, 'hh:mm:ss S')).to.be(`${`${date.getHours()}`.padStart(2, '0')}:${`${date.getMinutes()}`.padStart(2, '0')}:${`${date.getSeconds()}`.padStart(2, '0')} ${`${date.getMilliseconds()}`.padStart(3, '0')}`);
  });
});

describe('Formatter.money', () => {
  it('works fine', () => {
    expect(Formatter.money('100', ' ')).to.be('100');
    expect(Formatter.money('1000000000', ' ')).to.be('1 000 000 000');
    expect(Formatter.money('1000000000.999999', ' ')).to.be('1 000 000 000.999999');
  });

  it('works without delimiter', () => {
    expect(Formatter.money('100')).to.be('100');
    expect(Formatter.money('1000000000')).to.be('1 000 000 000');
    expect(Formatter.money('1000000000.999999')).to.be('1 000 000 000.999999');
  });

  it('works with custom delimiter', () => {
    expect(Formatter.money('100', ',')).to.be('100');
    expect(Formatter.money('1000000000', ',')).to.be('1,000,000,000');
    expect(Formatter.money('1000000000.999999', ',')).to.be('1,000,000,000.999999');
    expect(Formatter.money('1000000000', '')).to.be('1000000000');
  });

  it('works with fixed number', () => {
    expect(Formatter.money('100', ',', 0)).to.be('100');
    expect(Formatter.money('100', ',', 2)).to.be('100.00');
    expect(Formatter.money('1000000000.8888', ',', 2)).to.be('1,000,000,000.89');
    expect(Formatter.money('1000000000.99', ',', 5)).to.be('1,000,000,000.99000');
  });
});

describe('Formatter.cnmobile', () => {
  it('works fine', () => {
    expect(Formatter.cnmobile('13888888888', ' ')).to.be('1388 8888 888');
    expect(Formatter.cnmobile('+8613888888888', ' ')).to.be('+86 1388 8888 888');
    expect(Formatter.cnmobile('08613888888888', ' ')).to.be('086 1388 8888 888');
  });

  it('works without delimiter', () => {
    expect(Formatter.cnmobile('13888888888')).to.be('1388 8888 888');
    expect(Formatter.cnmobile('+8613888888888')).to.be('+86 1388 8888 888');
    expect(Formatter.cnmobile('08613888888888')).to.be('086 1388 8888 888');
  });

  it('works with custom delimiter', () => {
    expect(Formatter.cnmobile('13888888888', ',')).to.be('1388,8888,888');
    expect(Formatter.cnmobile('+8613888888888', ',')).to.be('+86,1388,8888,888');
    expect(Formatter.cnmobile('08613888888888', ',')).to.be('086,1388,8888,888');
  });
});

describe('Formatter.card', () => {
  it('works fine', () => {
    expect(Formatter.card('1234', ' ')).to.be('1234');
    expect(Formatter.card('12345678', ' ')).to.be('1234 5678');
    expect(Formatter.card('123456789012', ' ')).to.be('1234 5678 9012');
    expect(Formatter.card('1234567890123456', ' ')).to.be('1234 5678 9012 3456');
    expect(Formatter.card('1234567890123456789', ' ')).to.be('1234 5678 9012 3456 789');
  });

  it('works without delimiter', () => {
    expect(Formatter.card('1234')).to.be('1234');
    expect(Formatter.card('12345678')).to.be('1234 5678');
    expect(Formatter.card('123456789012')).to.be('1234 5678 9012');
    expect(Formatter.card('1234567890123456')).to.be('1234 5678 9012 3456');
    expect(Formatter.card('1234567890123456789')).to.be('1234 5678 9012 3456 789');
  });

  it('works with custom delimiter', () => {
    expect(Formatter.card('12345678', ',')).to.be('1234,5678');
    expect(Formatter.card('123456789012', ',')).to.be('1234,5678,9012');
    expect(Formatter.card('1234567890123456', ',')).to.be('1234,5678,9012,3456');
    expect(Formatter.card('1234567890123456789', ',')).to.be('1234,5678,9012,3456,789');
  });

  it('works with large number', () => {
    expect(Formatter.money('1111111111111111111', ',', 2)).to.be('1,111,111,111,111,111,111.00');
    expect(Formatter.money('11111111111111111111111111111111111111111111111111', ',', 2)).to.be('11,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111.00');
  });
});


src/Formatter.js
/* eslint no-console: "off" */
/**
 * Formatter Component for uxcore
 * @author eternalsky
 *
 * Copyright 2014-2015, Uxcore Team, Alinw.
 * All rights reserved.
 */

import Big from 'big.js';

const Formatter = {};

Formatter.date = (str, pattern) => {
  // new Date(null,false) 会返回最初日期。
  if (str === null || str === false || str === undefined || str === '') {
    console.warn('Formatter: invalid date');
    return '';
  }
  const date = new Date(str);
  if (Object.prototype.toString.call(date) === '[object Date]') {
    if (isNaN(date.getTime())) {
      // invalid
      console.warn('Formatter: invalid date');
      return '';
    }
    let actualPattern = pattern || 'YYYY-MM-DD';
    const o = {
      'M+': date.getMonth() + 1, // 月份
      'D+': date.getDate(), // 日
      'd+': date.getDate(), // 日
      'H+': date.getHours(), // 小时
      'h+': date.getHours(), // 小时
      'm+': date.getMinutes(), // 分
      's+': date.getSeconds(), // 秒
      'Q+': Math.floor((date.getMonth() + 3) / 3), // 季度
      'q+': Math.floor((date.getMonth() + 3) / 3), // 季度
      S: date.getMilliseconds(), // 毫秒
    };
    if (/(y+)/i.test(actualPattern)) {
      actualPattern = actualPattern.replace(
        RegExp.$1,
        (`${date.getFullYear()}`).substr(4 - RegExp.$1.length)
      );
    }
    // for (const k in o) {
    //   if (new RegExp(`(${k})`).test(actualPattern)) {
    //     actualPattern = actualPattern.replace(RegExp.$1,
    //       (RegExp.$1.length === 1) ? (o[k]) : ((`00${o[k]}`).substr((`${o[k]}`).length))
    //     );
    //   }
    // }
    Object.keys(o).forEach((k) => {
      if (new RegExp(`(${k})`).test(actualPattern)) {
        actualPattern = actualPattern.replace(RegExp.$1,
          (RegExp.$1.length === 1) ? (o[k]) : ((`00${o[k]}`).substr((`${o[k]}`).length))
        );
      }
    });
    return actualPattern;
  }
  return '';
};

Formatter.money = (str, delimiter = ' ', fixedNum) => {
  // const actualStr = fixedNum ? parseFloat(str).toFixed(fixedNum).toString() : str;
  const actualStr = fixedNum ? new Big(str).toFixed(parseInt(fixedNum, 10)).toString() : str;
  if (actualStr.indexOf('.') !== -1) {
    return actualStr.replace(/(\d)(?=(?:\d{3})+(\.))/g, (match, $1) => $1 + delimiter)
      .replace(/(\d{3})(?![$|\.|\(|\s])/g, (match, $1) => $1);
  }
  return actualStr.replace(/(\d)(?=(?:\d{3})+$)/g, (match, $1) => $1 + delimiter);
};

Formatter.cnmobile = (str, delimiter = ' ') => {
  return str.replace(/^(\+?0?86)(?!$)/,
    `$1${delimiter}`
  ).replace(/(\d{4})(?!$)/g, `$1${delimiter}`);
};

Formatter.card = (str, delimiter = ' ') => {
  return str.replace(/(\d{4})(?!$)/g, `$1${delimiter}`);
};

export default Formatter;


demo/FormatterDemo.js
/**
 * Formatter Component Demo for uxcore
 * @author guanghong.wsj
 *
 * Copyright 2014-2015, Uxcore Team, Alinw.
 * All rights reserved.
 */

const Formatter = require('../src');

const log = window.console.log;

log(Formatter.date(new Date(), 'YYYY-MM-DD HH:mm:ss')); // 2015-10-12 14:22:16
log(Formatter.money('6566456.65466545')); // "6 566 456.65466545"
log(Formatter.money('6566456.65466545', ',')); // "6,566,456.65466545"
log(Formatter.money('6566456', ',')); // "6,566,456"
log(Formatter.money('6566456', ',', 4)); // "6,566,456.0000"
log(Formatter.money('6566456', ',', 2)); // "6,566,456.00"
log(Formatter.money('6566456.65466545', ',', 4)); // "6,566,456.6547"
log(Formatter.money('6566456.65466545', ',', 2)); // "6,566,456.65"
log(Formatter.money('1111111111111111111', ',', 2)); // "1,111,111,111,111,111,111.00"
log(Formatter.money('11111111111111111111111111111111111111111111111111', ',', 2)); // "11,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111,111.00"

module.exports = Formatter;
